### ISSUE TITLE
Give guidance on how to migrate void-returning callback APIs to promise-returning

### ISSUE THREAD

**Issue by domenic** [w3ctag/promises-guide#24]
This will help spec authors looking to modernize.

**Comment by domenic** [comment]
Discussions with @ddorwin have made it clear this guide is massively failing to give good guidance on the type of things talked about in this thread. (See e.g. https://github.com/w3c/encrypted-media/issues/19.) I think we should prioritize giving clear patterns for:

1. Doing something after a promise callback triggers
2. Doing something before a promise callback triggers

I'll try and prioritize that.

**Comment by kenchris** [comment]
There is an newer comment from 24 February 2016: https://github.com/w3c/encrypted-media/issues/19#issuecomment-188497121

As I read this, the suggestion is that event handlers should run after promise fulfillment handlers (if additional microtasks (ie fulfilled promises) are run, they all run to completion before cycling the event loop.)

I also understand that this means calling in following in this order:
- resolve a promise (this automatically queues a task to resolve promise)
- queues a task to fire the event.

**Comment by annevk** [comment]
No, you want a single task to do both. I suspect that what happens these days is that the event is dispatched and promise resolving can be observed afterwards. So the event happens at an earlier point. I don't think anything went for my more complicated setup where the event would be dispatched as part of a "promise callback".

**Comment by kenchris** [comment]
I guess that is the advice we need to give

- if not on main thread (run in parallel), you must query a task for either dispatching an event or fulfilling a promise
- you should use the same task for both cases
- end result is that the promise fulfillment will run later than the event

**Comment by annevk** [comment]
Yeah. Also, there's no such thing is automatically queuing of tasks anymore. Someone could build such an abstraction again (and add it to HTML), but it would have to take more parameters and it's not clear we have that many consumers with such basic tasks that it's worth doing.
